src/uct/ib/ud/base/ud_ep.c

```diff
@@ -881,92 +1001,76 @@ static void uct_ud_ep_rx_ctl(uct_ud_iface_t *iface, uct_ud_ep_t *ep,
-    skb = uct_ud_iface_get_tx_skb(iface, ep);
-    if (!skb) {
-        return NULL;
-    }
-
-    neth = skb->neth;
-    uct_ud_neth_init_data(ep, neth);
-
-    neth->packet_type  = UCT_UD_EP_NULL_ID;
-    neth->packet_type |= UCT_UD_PACKET_FLAG_CTL;
-
-    creq = (uct_ud_ctl_hdr_t *)(neth + 1);
-
-    creq->type                = UCT_UD_PACKET_CREQ;
-    creq->conn_req.conn_sn    = ep->conn_sn;
-    creq->conn_req.path_index = ep->path_index;
+    uct_ud_ctl_hdr_t *ctl = (uct_ud_ctl_hdr_t*)(neth + 1);
 
-    status = uct_ud_ep_get_address(&ep->super.super,
-                                   (void*)&creq->conn_req.ep_addr);
-    if (status != UCS_OK) {
-        return NULL;
-    }
+    ucs_trace_func("");
 
-    status = uct_ib_iface_get_device_address(&iface->super.super.super,
-                                             (uct_device_addr_t*)uct_ud_creq_ib_addr(creq));
-    if (status != UCS_OK) {
-        return NULL;
+    switch (ctl->type) {
+    case UCT_UD_PACKET_CREQ:
+        break;
+    case UCT_UD_PACKET_CREP:
+        uct_ud_ep_rx_crep(iface, ep, neth, ctl);
+        break;
+    case UCT_UD_PACKET_FIN:
+        /* Got FIN packet which means that the endpoint won't be used to
+         * to data anymore */
+        ucs_assert(!(ep->flags & UCT_UD_EP_FLAG_CONNECT_TO_EP));
+        uct_ud_ep_handle_peer_disconnect(ep);
+        break;
+    default:
+        ucs_fatal("unexpected control packet type - %d", ctl->type);
     }
+}
 
-    uct_ud_peer_name(ucs_unaligned_ptr(&creq->peer));
+static UCS_F_NOINLINE void
+uct_ud_ep_process_unexp_packet(uct_ud_iface_t *iface, uct_ud_ep_t *ep,
+                               uct_ud_neth_t *neth, unsigned length)
+{
+    char buf[256] = {0};
 
-    skb->len = sizeof(*neth) + sizeof(*creq) + iface->super.addr_size;
-    return skb;
+    uct_ud_dump_packet(&iface->super.super, UCT_AM_TRACE_TYPE_RECV, neth,
+                       length, length, buf, sizeof(buf) - 1);
+    ucs_fatal("ep=%p: unexpected non-AM packet: [%s]", ep, buf);
 }
 
 void uct_ud_ep_process_rx(uct_ud_iface_t *iface, uct_ud_neth_t *neth, unsigned byte_len,
                           uct_ud_recv_skb_t *skb, int is_async)
 {
+    uct_ud_ep_t *ep = NULL;
     uint32_t dest_id;
     uint32_t is_am, am_id;
-    uct_ud_ep_t *ep = 0; /* todo: check why gcc complaints about uninitialized var */
     ucs_frag_list_ooo_type_t ooo_type;
 
-    UCT_UD_IFACE_HOOK_CALL_RX(iface, neth, byte_len);
+    UCT_UD_IFACE_HOOK_CALL_RX(iface, neth, goto out);
 
     dest_id = uct_ud_neth_get_dest_id(neth);
     am_id   = uct_ud_neth_get_am_id(neth);
     is_am   = neth->packet_type & UCT_UD_PACKET_FLAG_AM;
 
     if (ucs_unlikely(dest_id == UCT_UD_EP_NULL_ID)) {
-        /* must be connection request packet */
-        uct_ud_ep_rx_creq(iface, neth);
-        goto out;
+        /* Must be connection request packet */
+        ep = uct_ud_ep_rx_creq(iface, neth);
+        if (ep == NULL) {
+            goto out;
+        }
+
+        goto frag_list_insert;
```

Comment:
why needed? can't just goto out?
i don't think that when we failed to create the ep we want to add CREQ to frag_list

Reply:
at this point EP was created successfully and we should go to frag_list.
because this CREQ could be sent in the middle of the communication flow (not as a first packet).